app/routes/_protected.tsx

```diff
@@ -1,74 +1,169 @@
+  }
+
+  return user;
+};
+
+const routeGuardMiddleware: Route.unstable_ClientMiddlewareFunction = async (
+  { request },
+  next,
+) => {
+  console.debug('executing _protected.tsx route guard middleware');
+
+  const location = new URL(request.url);
+  // We have to use window.location.hash because location that comes from the request does not have the hash
+  const hash = window.location.hash;
+  const queryClient = getQueryClient();
+  const { isLoggedIn, user: authUser } = await queryClient.ensureQueryData(
+    authQuery(),
+  );
+  const shouldRedirectToSignup = !isLoggedIn;
+  const shouldVerifyEmail = authUser ? shouldUserVerifyEmail(authUser) : false;
   const isVerifyEmailRoute = location.pathname.startsWith('/verify-email');
   const shouldRedirectToVerifyEmail = shouldVerifyEmail && !isVerifyEmailRoute;
 
   console.debug('Rendering protected layout', {
+    time: new Date().toISOString(),
     location: location.pathname,
-    loading,
     isLoggedIn,
-    user,
     shouldRedirectToSignup,
+    userId: authUser?.id,
     shouldVerifyEmail,
     isVerifyEmailRoute,
     shouldRedirectToVerifyEmail,
   });
 
   if (shouldRedirectToSignup) {
-    return (
-      <Redirect
-        to={`/signup?redirectTo=${location.pathname}`}
-        logMessage="Redirecting from protected page to signup"
-      >
-        <LoadingScreen />
-      </Redirect>
-    );
+    let search = location.search;
+    if (location.pathname !== '/') {
+      const searchParams = new URLSearchParams(location.search);
+      searchParams.set('redirectTo', location.pathname);
+      search = `?${searchParams.toString()}`;
+    }
+
+    throw redirect(`/signup${search}${hash}`);
   }
 
+  await ensureUserData(queryClient, authUser);
+
   if (shouldRedirectToVerifyEmail) {
     const searchParams = new URLSearchParams(location.search);
-    searchParams.set('redirectTo', location.pathname);
-
-    return (
-      <Redirect
-        to={{
-          ...location,
-          pathname: '/verify-email',
-          search: searchParams.toString(),
-        }}
-        logMessage="Redirecting from protected page to verify email"
-      >
-        <LoadingScreen />
-      </Redirect>
-    );
+    if (location.pathname !== '/') {
+      searchParams.set('redirectTo', location.pathname);
+    }
+    const search = `?${searchParams.toString()}`;
+
+    throw redirect(`/verify-email${search}${hash}`);
   }
 
-  if (!isLoggedIn) {
-    return <LoadingScreen />;
+  await next();
+};
+
+export const unstable_clientMiddleware: Route.unstable_ClientMiddlewareFunction[] =
+  [routeGuardMiddleware];
+
+export async function clientLoader() {
+  // We are keeping this clientLoader to force client rendering for all protected routes.
+}
+
+clientLoader.hydrate = true as const;
+
+export function HydrateFallback() {
+  return <LoadingScreen />;
+}
+
+export default function ProtectedRoute() {
+  const { user } = useAuthState();
+
+  if (!user) {
+    console.debug('Logging out...');
```

Comment:
This can happen theoretically while logout is in progress but I almost never see this log